ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStoreV2.java

```diff
@@ -63,10 +78,100 @@ public CloudBlobStoreV2(VerifiableProperties properties, MetricRegistry metricRe
+   * It fails for any other exception.
+   * @param messageSetToWrite The message set to write to the store
+   *                          Only the StoreKey, OperationTime, ExpirationTime, LifeVersion should be used in this method.
+   * @throws StoreException
+   */
   @Override
   public void put(MessageWriteSet messageSetToWrite) throws StoreException {
-    // TODO
-  }
+    MessageFormatWriteSet messageFormatWriteSet = (MessageFormatWriteSet) messageSetToWrite;
+    Timer.Context storageTimer = null;
+    // For-each loop must be outside try-catch. Loop must continue on BLOB_ALREADY_EXISTS exception
+    for (MessageInfo messageInfo : messageFormatWriteSet.getMessageSetInfo()) {
+      try {
+        // Read blob from input stream. It is not possible to just pass the input stream to azure-SDK.
+        // The SDK expects the stream to contain exactly the number of bytes to be read. If there are
+        // more bytes, it fails assuming the stream is corrupted. The inputStream we have here is a
+        // concatenation of many blobs and not just one blob.
+        ByteBuffer messageBuf = ByteBuffer.allocate((int) messageInfo.getSize());
+        if (messageFormatWriteSet.getStreamToWrite().read(messageBuf.array()) == -1) {
+          throw new RuntimeException(
+              String.format("Failed to read blob %s with %s bytes because end-of-stream was reached",
+                  messageInfo.getStoreKey().getID(), messageInfo.getSize()));
+        }
+
+        // Prepare to upload blob to Azure blob storage
+        // There is no parallelism, but we still need to create and pass this object to SDK.
+        BlobParallelUploadOptions blobParallelUploadOptions =
+            new BlobParallelUploadOptions(new ByteBufferInputStream(messageBuf));
+        // To avoid overwriting, pass "*" to setIfNoneMatch(String ifNoneMatch)
+        // https://learn.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable
+        blobParallelUploadOptions.setRequestConditions(new BlobRequestConditions().setIfNoneMatch("*"));
+        blobParallelUploadOptions.setMetadata(messageInfoToMap(messageInfo));
+        // Without content-type, get-blob floods log with warnings
+        blobParallelUploadOptions.setHeaders(new BlobHttpHeaders().setContentType("application/octet-stream"));
+
+        ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
+        storageTimer = azureMetrics.blobUploadTime.time();
+        Response<BlockBlobItem> blockBlobItemResponse =
+            blobContainerClient.getBlobClient(messageInfo.getStoreKey().getID())
+                .uploadWithResponse(blobParallelUploadOptions, Duration.ofMillis(cloudConfig.cloudRequestTimeout),
+                    Context.NONE);
+        ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
```

Comment:
Do we still support encryption?

Reply:
V1 does not have encryption as `vcrRequireEncryption = false` on all VCR clusters.

Reply:
Ack.